### PR TITLE
ingress syncer agent

### PR DESCRIPTION
**Pull request by philbrookes**
agent running on workload clusters, watches ingress and annotates DNS records in the control plane with required changes.

verification:
```
make local-setup MCTC_WORKLOAD_CLUSTERS_COUNT=1
```

Update the `config/manager/manager.yaml` to contain the necessary environment variable and update the image policy (N.B. Do not copy/paste this, the hyphens from github are bad for the YAML parser):
```
    containers:
        - command:
            - /controller
            args:
            - --leader-elect
            image: controller:latest
            imagePullPolicy: Never
            env:
            - name: AWS_ACCESS_KEY_ID
            value: <AWS_ACCESS_KEY_ID>
            - name: AWS_SECRET_ACCESS_KEY
            value: <AWS_SECRET_ACCESS_KEY>
            - name: AWS_DNS_PUBLIC_ZONE_ID
            value: <AWS_DNS_PUBLIC_ZONE_ID>
            - name: ZONE_ROOT_DOMAIN
            value: <ZONE_ROOT_DOMAIN>
``` 
Load the image into kind:
```
export KUBECONFIG=./hack/kubeconfigs/mctc-control-plane.kubeconfig
make kind-load-controller
```
deploy the image:
```
make deploy-controller
```
Target the workload cluster you wish to run on:
```sh
export KUBECONFIG=./hack/kubeconfigs/mctc-workload-1.kubeconfig
```
Create the control plane secret
Update the `samples/syncer/secret.yaml` from the `hack/kubeconfigs/mctc-control-plane.kubeconfig` and apply it:
```sh
kubectl create -f samples/syncer/secret.yaml
```
Run the agent locally:
```sh
make run-agent
```
Create the sample application (in a second terminal also targetting the workload cluster):
```sh
export KUBECONFIG=./hack/kubeconfigs/mctc-workload-1.kubeconfig
kubectl apply -f ./samples/syncer/application.yaml
```
Wait for  the agent to log:
```
certificate secret in place for  host adding dns endpoints	{"host": "s6yxqyfx3awf53iasbk7a6.pb.hcpapps.net"}
```
Then after a few moments (for propagation) dig the host:
```
dig s6yxqyfx3awf53iasbk7a6.pb.hcpapps.net
```
once it is resolving, curl it:
```
curl -k https://s6yxqyfx3awf53iasbk7a6.pb.hcpapps.net
```

### CODE REVIEW

**Comment by maleck13** [comment]
is this covering the first three points here: https://github.com/orgs/Kuadrant/projects/9/views/4?pane=issue&itemId=19502318 ?

**Comment by philbrookes** [comment]
> is this covering the first three points here: https://github.com/orgs/Kuadrant/projects/9/views/4?pane=issue&itemId=19502318 ?

Yes.

File: Dockerfile [hunk 1]
```diff
@@ -12,21 +12,31 @@ COPY go.sum go.sum
...
 COPY pkg/ pkg/
 
 # Build
 # the GOARCH has not a default value to allow the binary be built according to the host where the command
 # was called. For example, if we call make docker-build in a local env which has the Apple Silicon M1 SO
 # the docker BUILDPLATFORM arg will be linux/arm64 when for Apple x86 it will be linux/amd64. Therefore,
 # by leaving it empty we can ensure that the container and binary shipped on it will have the same platform.
-RUN CGO_ENABLED=0 GOOS=${TARGETOS:-linux} GOARCH=${TARGETARCH} go build -a -o manager main.go
+RUN CGO_ENABLED=0 GOOS=${TARGETOS:-linux} GOARCH=${TARGETARCH} go build -a -o controller cmd/controller/main.go
+RUN CGO_ENABLED=0 GOOS=${TARGETOS:-linux} GOARCH=${TARGETARCH} go build -a -o agent cmd/agent/main.go
```
Comment (sergioifg94) [inline Dockerfile:24]: We will want to update this function that checks whether the controller is running locally or as a container. It does it by checking that the `/manager` binary exists, so now that it has been renamed it won't work

https://github.com/Kuadrant/multi-cluster-traffic-controller/blob/main/pkg/_internal/controller/run.go#L13

File: Makefile [hunk 2]
```diff
@@ -1,6 +1,16 @@
 
 # Image URL to use all building/pushing image targets
-IMG ?= controller:latest
+TAG ?= latest
+CONTROLLER_IMG ?= pbrookes/controller:$(TAG)
```
Comment (maleck13) [inline Makefile:4]: could this not be your specific image?

File: Dockerfile (hunk 1 above)
Comment (philbrookes) [reply Dockerfile:24]: thanks Sergio, updated this now.

File: Makefile (hunk 2 above)
Comment (philbrookes) [reply Makefile]: ah, overlooked those ones, updated now.

**Comment by philbrookes** [comment]
Moved the PR: https://github.com/Kuadrant/multi-cluster-traffic-controller/pull/70

**Pull request closed without merging**